python/ray/serve/_private/http_proxy.py

```diff
@@ -1148,28 +1150,72 @@ async def _consume_and_send_asgi_message_generator(
+                    obj_ref_generator._next_async(
+                        timeout_s=calculate_remaining_timeout(
+                            timeout_s=timeout_s,
+                            start_time_s=start,
+                            curr_time_s=time.time(),
+                        )
                     )
                 )
+                tasks = [next_obj_ref_task]
+                # Once the response is completed, the client will immediately
+                # disconnect. This causes a race condition: if we get the disconnect
+                # message before the replica handler finishes, we'll falsely detect a
+                # client disconnect. To avoid this, stop listening for disconnects once
+                # the full response has been sent.
+                if not response_done:
+                    tasks.append(disconnected_task)
+
+                done, _ = await asyncio.wait(
+                    tasks,
+                    return_when=asyncio.FIRST_COMPLETED,
+                    # No timeout is set because it's already passed to `_next_async`.
+                )
+                if disconnected_task in done:
+                    logger.info(
+                        "Client disconnected during execution, cancelling request."
+                    )
+                    next_obj_ref_task.cancel()
+                    ray.cancel(obj_ref_generator)
+                    status_code = DISCONNECT_ERROR_CODE
+                    break
+
+                obj_ref = next_obj_ref_task.result()
                 if obj_ref.is_nil():
                     raise RayServeTimeout(is_first_message=is_first_message)
 
                 asgi_messages: List[Message] = pickle.loads(await obj_ref)
+                # See the ASGI spec for message details:
+                # https://asgi.readthedocs.io/en/latest/specs/www.html.
                 for asgi_message in asgi_messages:
                     if asgi_message["type"] == "http.response.start":
                         # HTTP responses begin with exactly one
                         # "http.response.start" message containing the "status"
-                        # field Other response types (e.g., WebSockets) may not.
+                        # field. Other response types (e.g., WebSockets) may not.
                         status_code = str(asgi_message["status"])
+                        expecting_trailers = asgi_message.get("trailers", False)
+                    elif (
+                        asgi_message["type"] == "http.response.body"
+                        and not asgi_message.get("more_body", False)
+                        and not expecting_trailers
+                    ):
+                        # If the body is completed and we aren't expecting trailers, the
+                        # response is done.
+                        response_done = True
+                    elif asgi_message["type"] == "http.response.trailers":
```

Comment:
There's [a `more_trailers` flag](https://asgi.readthedocs.io/en/stable/extensions.html#http-trailers) in `trailer` responses. Should we check whether that's set to `False` in this conditional to make sure there's no more `trailer` responses?

Reply:
oh great catch. yes, i've added this